src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -2554,6 +2627,319 @@ public final void addBatch() throws SQLServerException {
+            // If it's followed by a dot, then it's a database.
+            // Otherwise, it's the table.
+            if (localUserSQL.charAt(tempint + 1) == '.') {
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 2);
+                return tempstr + "." + parseUserSQLForTableNameDW(hasInsertBeenFound, hasIntoBeenFound);
+            } else {
+                // return tablename
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 1);
+                return tempstr;
+            }
+        }
+        
+        // do the same for ""
+        if (localUserSQL.substring(0, 1).equalsIgnoreCase("\"")) {
+            int tempint = localUserSQL.indexOf("\"", 1);
+            
+            // keep checking if it's escaped
+            while (localUserSQL.charAt(tempint + 1) == '\"') {
+                localUserSQL = localUserSQL.substring(0, tempint) + localUserSQL.substring(tempint + 1);
+                tempint = localUserSQL.indexOf("\"", tempint + 1);
+            }
+            
+            // we've found a " that is actually trying to close the quote.
+            // If it's followed by a dot, then it's a database.
+            // Otherwise, it's the table.
+            if (localUserSQL.charAt(tempint + 1) == '.') {
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 2);
+                return tempstr + "." + parseUserSQLForTableNameDW(hasInsertBeenFound, hasIntoBeenFound);
+            } else {
+                // return tablename
+                String tempstr = localUserSQL.substring(1, tempint);
+                localUserSQL = localUserSQL.substring(tempint + 1);
+                return tempstr;
+            }
+        }
+        
+        // At this point, the next chunk of string is the table name (could have database name), without starting with [ or ".
+        StringBuilder sb = new StringBuilder();
+        while (localUserSQL.length() > 0) {
+            if (localUserSQL.charAt(0) == '.' || localUserSQL.charAt(0) == ' ' || localUserSQL.charAt(0) == '\t'
+                    || localUserSQL.charAt(0) == '(') {
+                if (localUserSQL.charAt(0) == '.') {
+                    localUserSQL = localUserSQL.substring(1);
+                    return sb.toString() + "." + parseUserSQLForTableNameDW(hasInsertBeenFound, hasIntoBeenFound);
+                } else {
+                    return sb.toString();
+                }
+            } else {
+                sb.append(localUserSQL.charAt(0));
+                localUserSQL = localUserSQL.substring(1);
+            }
+        }
+        
+        // It shouldn't come here. If we did, something is wrong.
+        throw new IllegalArgumentException("localUserSQL");
```

Comment:
Why did you pick throwing an IllegalArgumentException (for this & the other parse routines)?  It's probably better to have a descriptive error message instead.

Reply:
I decided to throw an IllegalArgumentException because I thought it best described the symptom. It would be easy to throw an exception detailing at which index the parsing failed if the parsing actually failed at some point, but for these instances where the parsing fails due to getting to the end of the parsing method (which it shouldn't - this means the user's SQL query was incorrect somehow), all we know is that the user's argument was incorrect. I do agree that I could give a better description, though - I'll add that in.